feat(launcher): Provide option to run puppeteer with different browsers

mjzffr opened the pull request:
This is an initial refactoring of Launcher to enable run-time browser selection in Puppeteer, starting with Firefox. On our side, we're tracking this and related work at https://bugzilla.mozilla.org/show_bug.cgi?id=1590467

/cc @whimboo @andreastt @AutomatedTester

mjzffr committed:
feat: Set which browser to launch via PUPPETEER_PRODUCT

This change introduces a PUPPETEER_PRODUCT environment
variable as a first step toward using Puppeteer with
many different browsers. Setting PUPPETEER_PRODUCT=firefox, for
example, enables Firefox-specific Launcher settings.

The state is also exposed as `puppeteer.product` in the API
to support adding other product-specific behaviour as needed.

The bulk of the change is a refactoring in Launcher
to decouple generic browser start-up from product-specific
configuration.

mjzffr committed:
test: Distinguish Juggler unit tests from Firefox

The funit script is renamed to fjunit (j for Juggler, which is
used only by the experimental puppeteer-firefox package.

In contrast, the funit script now refers to running Puppeteer
unit tests against the main puppeteer package with Firefox.
To do so with Firefox Nightly, run:

`BINARY=path/to/firefox npm run funit`

A number of changes in this patch make it easier to run
Puppeteer unit tests in Mozilla's CI.

mjzffr committed:
fix(helper): simplify emptying of array

mjzffr commented:
@mathiasbynens Hi! This is ready for review. My CLA is in the works.

I looked at the results of the checks and it seems the failures are known to be intermittent... Let me know.

mjzffr commented:
@googlebot I signed it!

mathiasbynens commented:
Yay, thanks for sending this over! Just as a heads-up, I won’t be able to review this as soon as I’d like, since I’m traveling for ChromeDevSummit and BlinkOn this week. I’ll get back to y’all as soon as I can.

mathiasbynens commented on test/puppeteer.spec.js:
```diff
@@ -28,31 +28,40 @@ module.exports.addTests = ({testRunner, product, puppeteerPath}) => {
-    // Make sure the `npm install` was run after the chromium roll.
-    if (!fs.existsSync(puppeteer.executablePath()))
-      throw new Error(`Browser is not downloaded. Run 'npm install' and try to re-run tests`);
+    const path = puppeteer.executablePath();
```
Let’s use a different identifier here so as not to shadow `path` (so that it `path` remains `require('path')` throughout this entire file). How about `executablePath`?

mjzffr replied:
Done.

mathiasbynens commented on lib/helper.js:
```diff
@@ -137,7 +137,7 @@ class Helper {
     for (const listener of listeners)
       listener.emitter.removeListener(listener.eventName, listener.handler);
-    listeners.splice(0, listeners.length);
+    listeners.length = 0;
```
Ooh, drive-by simplification! I dig it 👍🏻

mathiasbynens commented on test/puppeteer.spec.js:
```diff
@@ -28,31 +28,40 @@ module.exports.addTests = ({testRunner, product, puppeteerPath}) => {
+  let extraLaunchOptions = {};
+  try {
+    extraLaunchOptions = JSON.parse(process.env.EXTRA_LAUNCH_OPTIONS || '{}');
+  } catch (e) {
```
nit: `e` → `error` or `exception`

mathiasbynens commented on lib/Puppeteer.js:
```diff
@@ -18,6 +18,8 @@ const BrowserFetcher = require('./BrowserFetcher');
 const Errors = require('./Errors');
 const DeviceDescriptors = require('./DeviceDescriptors');
 
+const PRODUCT = process.env.PUPPETEER_PRODUCT || process.env.npm_config_puppeteer_product || process.env.npm_package_config_puppeteer_product;
```
Let’s move these `env` variables elsewhere — `puppeteer-core` must remain agnostic, and only `lib/Launcher.js` and `install.js` should consider them. `lib/Launcher.js` seems like a good fit. WDYT?

mjzffr replied:
Yes, there may be a better place for this.

The main motivation here is to maintain some sort of privileged state about which browser we're talking to. This is both for Puppeteer API consumers and the Puppeteer implementation itself, since both internally and externally some browser-specific code may be needed from time to time.

The same holds true even in the case of `puppeteer-core`: the library doesn't download any browser binary for us, but which browser is it ultimately talking to? Or am I mistaken?

Given that, do you have any suggestions?

mjzffr replied:
^ @mathiasbynens 

Mentioning you in case this question got lost in a sea of notifications. :)

mathiasbynens replied:
The problem is that `puppeteer-core` might be run outside of Node.js, where `process.*` is not available. In the Launcher, we wrap any usage of `process.env` in `if (!this._isPuppeteerCore) { ... }` for that reason.

mathiasbynens replied:
Maybe I'm misunderstanding. What problems do you see with keeping the logic local to `lib/Launcher.js`?

mjzffr replied:
I think I've addressed both our concerns with the latest change. :) The `PUPPETEER_PRODUCT` stuff is only used if we're not running `puppeteer-core`, and I changed Puppeteer to lazily load an appropriate Launcher based on a `product` option, if available, falling back on environment variables.

mathiasbynens commented on lib/Launcher.js:
```diff
@@ -294,49 +342,193 @@ class Launcher {
+   */
+  async _createProfile(extraPrefs) {
+    const profilePath = await mkdtempAsync(path.join(os.tmpdir(), 'puppeteer_dev_firefox_profile-'));
+    const prefs_js = [];
```
coding style nit: `prefs_js` → `prefsJs`

same for `user_js`

mathiasbynens commented on lib/Launcher.js:
```diff
@@ -294,49 +342,193 @@ class Launcher {
+function waitForWSEndpoint(browserProcess, timeout, preferredRevision) {
   return new Promise((resolve, reject) => {
-    const rl = readline.createInterface({ input: chromeProcess.stderr });
+    const rl = readline.createInterface({ input: browserProcess.stderr});
```
nit: space before `}`

mathiasbynens reviewed (changes requested):
I left some initial comments and questions. I want to do a more thorough second pass after BlinkOn, but figured this should get the conversation started.

Thanks again for your work on this!

mjzffr committed:
(lib) fixups

mjzffr committed:
(test) fixups

mathiasbynens commented on docs/api.md:
```diff
@@ -547,6 +549,7 @@ try {
   - `env` <[Object]> Specify environment variables that will be visible to the browser. Defaults to `process.env`.
   - `devtools` <[boolean]> Whether to auto-open a DevTools panel for each tab. If this option is `true`, the `headless` option will be set `false`.
   - `pipe` <[boolean]> Connects to the browser over a pipe instead of a WebSocket. Defaults to `false`.
+  - `extraPrefs` <[Object]> Additional [preferences](https://developer.mozilla.org/en-US/docs/Mozilla/Preferences/Preference_reference) that can be passed to Firefox (see `PUPPETEER_PRODUCT`)
```
If we're gonna add a Firefox-specific option, it should probably have Firefox in its name, no?

mjzffr replied:
Sure. Done.

mathiasbynens commented on lib/Launcher.js:
```diff
@@ -412,6 +604,52 @@ function getWSEndpoint(browserURL) {
+ * @param {boolean} isPuppeteerCore
+ * @return {ChromeLauncher|FirefoxLauncher}
+ */
+function Launcher(product, projectRoot, preferredRevision, isPuppeteerCore) {
```
Let's rename this `launch` now that it's no longer a class

mjzffr replied:
I'm happy to, but it implies a load of type annotation changes (e.g. `Launcher.BrowserOptions`), and I'm not sure that's what you want. 

I left it as `Launcher`, but do let me know if you still want `launcher`.

mjzffr committed:
(fixup) Enable setting a product programmatically

Respecting the puppeteer-core restriction for PUPPETEER_
environment variables, lazily instantiate the Launcher
based on a `product` Puppeteer.launch option, if available.

hanselfmu reviewed (approved):
LGTM.

rodoabad commented:
Have the documents been updated on how to use this new feature? i.e. do you still need `puppeteer-firefox`?

chrisLovesCode commented:
> Have the documents been updated on how to use this new feature? i.e. do you still need `puppeteer-firefox`?

`puppeteer.launch({product: 'firefox'})`
but it's classified as experimental.
Have a look at the FAQ here: https://github.com/puppeteer/puppeteer#faq

rodoabad commented:
@chrisLovesCode it doesn't say if `puppeteer` downloads firefox like how it downloads chromium or not. If we still need `puppeteer-firefox` then that's fine. Just wish the documentation was clear on that apart from explaining the new API.

chrisLovesCode commented:
okay , i got your point! ..yes i agree, would be nice to get some informations, currently i  have not tried out to use firefox, but would like to do it soon, since i will write e2e tests with puppeteer

mjzffr commented:
@rodoabad puppeteer doesn't download Firefox. That's tracked in https://github.com/puppeteer/puppeteer/issues/5151 

Totally agree that this should be clearer from the docs; I expect that will get better as progress is made on Firefox support.
